### PR TITLE
Add notification metrics

### PR DESCRIPTION
**Pull request by tifayuki**
It adds notification related prometheus metrics, including:
  - total count for events/success/failure/error
  - total count for notification per each status code
  - gauge of the pending notification queue

Signed-off-by: tifayuki <tifayuki@gmail.com>

### CODE REVIEW

**Comment by dmp42** [comment]
Same as the other PR.
Can we see the output?

Thanks.

File: metrics/prometheus.go [hunk 1]
```diff
@@ -10,4 +10,7 @@ const (
 var (
 	// StorageNamespace is the prometheus namespace of blob/cache related operations
 	StorageNamespace = metrics.NewNamespace(NamespacePrefix, "storage", nil)
+
+	// NotificationNamespace is the prometheus namespace of notification related metrics
+	NotificationNamespace = metrics.NewNamespace(NamespacePrefix, "notification", nil)
```
Comment (stevvooe) [inline metrics/prometheus.go]: Pluralize this as `notifications`.

**Comment by tifayuki** [comment]
An sample output is:
```
# HELP registry_notifications_events_total The number of total events
# TYPE registry_notifications_events_total counter
registry_notifications_events_total{type="Events"} 8
registry_notifications_events_total{type="Successes"} 8
# HELP registry_notifications_pending_total The gauge of pending events in queue
# TYPE registry_notifications_pending_total gauge
registry_notifications_pending_total 0
# HELP registry_notifications_status_total The number of status code
# TYPE registry_notifications_status_total counter
registry_notifications_status_total{code="200 OK"} 8
```
@stevvooe @dmp42

**Comment by mrueg** [comment]
@stevvooe any chance to take a look at this? would be great to have this change in.

**Comment by mrueg** [comment]
@dmcgowan @stevvooe  friendly ping on this one :)

**Comment by manishtomar** [comment]
@tifayuki Could you add endpoint label also to this PR?

**Comment by tifayuki** [comment]
@manishtomar I added the endpoint label. Can you take a look at it?

**Comment by manishtomar** [comment]
@tifayuki Thanks! Code LGTM. Could you merge latest master and push it? I wonder if it will fix travis build.

**Comment by tifayuki** [comment]
@manishtomar 
I merged the upstream master to this PR. But on Travis, it always shows: 
```
The command "export CGO_ENABLED=$TRAVIS_CGO_ENABLED" exited with 0.
0.23s$ DCO_VERBOSITY=-q script/validate/dco
/home/travis/gopath/bin/git-validation
 99ce26e - FAIL - has whitespace errors. See `git show --check 99ce26efeda8fc8661974a1d3edaab0740c2abc3`.
1 commits to fix
The command "DCO_VERBOSITY=-q script/validate/dco" exited with 1.
```
Any idea what error it is?

**Comment by mrueg** [comment]
@tifayuki can you rebase on master and see if the build still fails?

**Comment by tifayuki** [comment]
@mrueg it worked

**Comment by mrueg** [comment]
@dmcgowan any chance to get this in for 2.7.1?

**Comment by manishtomar** [comment]
@dmcgowan I added this in 2.7.2 milestone. Any chance this can be merged?

**Comment by dmcgowan** [comment]
LGTM

If you are going to open up a backport to 2.7 for this, please provide strong justification and show how it is not impactful for compatibility. Such a change in a patch release is very irregular, but I understand there is no 2.8 currently on the horizon.

**Comment by dmcgowan** [comment]
Can you squash this before merging? Will make cherry-picking cleaner

File: notifications/metrics.go [hunk 2]
```diff
@@ -61,19 +75,27 @@ func (emsl *endpointMetricsHTTPStatusListener) success(status int, events ...Eve
 	defer emsl.safeMetrics.Unlock()
 	emsl.Statuses[fmt.Sprintf("%d %s", status, http.StatusText(status))] += len(events)
 	emsl.Successes += len(events)
+
+	statusCounter.WithValues(fmt.Sprintf("%d %s", status, http.StatusText(status)), emsl.EndpointName).Inc(1)
+	eventsCounter.WithValues("Successes", emsl.EndpointName).Inc(1)
```
Comment (manishtomar) [inline notifications/metrics.go]:
```suggestion
	eventsCounter.WithValues("Successes", emsl.EndpointName).Inc(len(events))
```

File: notifications/metrics.go [hunk 3]
```diff
@@ -61,19 +75,27 @@ func (emsl *endpointMetricsHTTPStatusListener) success(status int, events ...Eve
...
 }
 
 func (emsl *endpointMetricsHTTPStatusListener) failure(status int, events ...Event) {
 	emsl.safeMetrics.Lock()
 	defer emsl.safeMetrics.Unlock()
 	emsl.Statuses[fmt.Sprintf("%d %s", status, http.StatusText(status))] += len(events)
 	emsl.Failures += len(events)
+
+	statusCounter.WithValues(fmt.Sprintf("%d %s", status, http.StatusText(status)), emsl.EndpointName).Inc(1)
+	eventsCounter.WithValues("Failures", emsl.EndpointName).Inc(1)
```
Comment (manishtomar) [inline notifications/metrics.go]:
```suggestion
	eventsCounter.WithValues("Failures", emsl.EndpointName).Inc(len(events))
```

File: notifications/metrics.go [hunk 4]
```diff
@@ -61,19 +75,27 @@ func (emsl *endpointMetricsHTTPStatusListener) success(status int, events ...Eve
...
 }
 
 func (emsl *endpointMetricsHTTPStatusListener) err(err error, events ...Event) {
 	emsl.safeMetrics.Lock()
 	defer emsl.safeMetrics.Unlock()
 	emsl.Errors += len(events)
+
+	eventsCounter.WithValues("Errors", emsl.EndpointName).Inc(1)
```
Comment (manishtomar) [inline notifications/metrics.go]:
```suggestion
	eventsCounter.WithValues("Errors", emsl.EndpointName).Inc(len(events))
```

File: notifications/metrics.go [hunk 5]
```diff
@@ -87,12 +109,17 @@ func (eqc *endpointMetricsEventQueueListener) ingress(events ...Event) {
 	defer eqc.Unlock()
 	eqc.Events += len(events)
 	eqc.Pending += len(events)
+
+	eventsCounter.WithValues("Events", eqc.EndpointName).Inc()
+	pendingGauge.WithValues(eqc.EndpointName).Inc(1)
```
Comment (manishtomar) [inline notifications/metrics.go]:
```suggestion
	pendingGauge.WithValues(eqc.EndpointName).Inc(len(events))
```

**Comment by manishtomar** [comment]
> If you are going to open up a backport to 2.7 for this, please provide strong justification and show how it is not impactful for compatibility.

It does not impact compatibility. I originally wanted it in 2.7 release as this change is being used in Hub registry but missed the train. I thought we could get it in 2.7.1|2 release but I that was assuming that 2.7.x release will be based on master branch which is incorrect. Hence, please ignore my request to put this change in patch release. It can go in next release. I just wanted to get this merged. Also I apologize I just realized a mistake when looking at the code and have requested these. We can merge it after @tifayuki makes those changes. Thanks for your time!

**Comment by manishtomar** [comment]
@tifayuki Thanks for the changes. However build is failing due to unsigned commits which was GH fault when accepting the suggestion. Could you please rebase with one signed commit? Also, you will need to put `.Inc(float64(len(events)))` to fix the lint error.

File: notifications/metrics.go [hunk 6]
```diff
@@ -5,6 +5,18 @@ import (
...
 	"net/http"
 	"sync"
+
+	prometheus "github.com/docker/distribution/metrics"
+	"github.com/docker/go-metrics"
+)
+
+var (
+	// eventsCounter counts total events of incoming, success, failure, and errors
+	eventsCounter = prometheus.NotificationsNamespace.NewLabeledCounter("events", "The number of total events", "type", "to")
```
Comment (manishtomar) [inline notifications/metrics.go]: Non-blocking nit: How about changing label name from `to` to `endpoint` instead? That terminology is consistent with codebase and config. Sorry I should've realized this before suggesting the name long time ago.

**Comment by manishtomar** [comment]
@dmcgowan Can this be merged now please?

**Comment by thaJeztah** [comment]
do we want some of the commits to be squashed before merging?

**Comment by manishtomar** [comment]
@tifayuki Can you please squash commits?

**Comment by manishtomar** [comment]
Actually since this only needs to be merged and not cherry-picked it should be ok to squash during merge @dmcgowan . If you can you or @caervs please merge this?

**Comment by manishtomar** [comment]
@caervs Can you please merge this?

**Pull request merged**